nav2_behavior_tree/plugins/action/truncate_path_local_action.cpp

```diff
@@ -0,0 +1,147 @@
+// Copyright (c) 2021 RoboTech Vision
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <limits>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "behaviortree_cpp_v3/decorator_node.h"
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_util/robot_utils.hpp"
+#include "nav_msgs/msg/path.hpp"
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_behavior_tree/plugins/action/truncate_path_local_action.hpp"
+
+namespace nav2_behavior_tree
+{
+
+TruncatePathLocal::TruncatePathLocal(
```

Comment:
Does this need to be in this same PR? It would be good to discuss this in another PR just to keep all of this focused on the smoother. I'm not 100% sure why this is needed, but it would be good to review / discuss the need separately

Reply:
It can be discussed after closing the smoother reviews (although I would keep it in the same PR if it's not a problem - that would be less work for me). Just shortly about the need: the smoother really needs an upper bound for the path length to provide good results in a reasonable time and this simple BT node takes care of that.

Reply:
I figured that would be something we could add to the SmoothPath action API to add the max length to send to the smoother. That seems like a more natural fit, no?

Reply:
> On the parameters, the optimization_length and optimization_length_backwards should probably be parameters of your smoother plugin rather than attributes of the server (since not every smoother will have that option / make use of it).

I'm sorry but we have already discussed that and these are your words from Slack. You asked me to remove the max length from the smoother server and we agreed upon adding the truncation BT node.

In addition, what profit would be from gluing two independent useful lego parts into one? The node can be used in BT for various other tasks. Actually we use it in our robots for further collision checking in front of robot, i.e. something like this:

```
                <TruncatePathLocal input_path="{path_coarse}" output_path="{path_coarse_local}"
                  distance_forward="8.0" distance_backward="4.0" robot_frame="base_footprint"/>
                <SmoothPath unsmoothed_path="{path_coarse_local}" smoothed_path="{smoothed_path}" smoother_id="SmoothPath" check_for_collisions="false"/>
                <TruncatePathLocal input_path="{smoothed_path}" output_path="{smoothed_path_for_check}"
                  distance_forward="3.5" distance_backward="0.0" robot_frame="base_footprint"/>
                <CheckPathValidity path_in="{smoothed_path_for_check}"/>
```

Note: SmoothPath with a VoidSmoother plugin (doing nothing with the path) and `check_for_collisions="true"`  can be used instead of CheckPathValidity (the name was used just for readability)

Reply:
Fair enough, I'll go with you on that then